Update usage example for DrawQuad - Beyond SK

sam-stajnko opened the pull request:
Splashkit Function: DrawQuad
Description: This is a Beyond Splashkit usage example for the function DrawQuad. There is already a usage example for this function using the Splashkit core library that draws a ninja star on the window using diamond shaped quads. The purpose of this pull request is to recreate the same usage example but only using SDL2 functions, without any reference to the Splashkit core library.

# Checklist
- [x] Beyond Splashkit C++ code
- [x] Rename previous C++ file
- [x] Code runs successfully without errors

prb616 commented on public/usage-examples/graphics/draw_quad-1-example-beyond.cpp:
```diff
@@ -0,0 +1,101 @@
+    Uint8 blue;
+};
+
+void draw_quad_rgb(SDL_Renderer * renderer, Point_2D pt_1, Point_2D pt_2, Point_2D pt_3, Point_2D pt_4, Color_RGB color)
```
you can create Quad struct instead of draw_quad_rgb() call manually passed 4 points + color. this will groups all relevant data for one shape together, making it easier to manage and reuse.

suggested code changes:

`struct Quad {
    Point_2D p1, p2, p3, p4;
    Color_RGB color;
};
`

prb616 commented on public/usage-examples/graphics/draw_quad-1-example-beyond.cpp:
```diff
@@ -0,0 +1,101 @@
+    SDL_RenderClear(renderer);
+
+    // Draw a ninja star using four quads
+    draw_quad_rgb(renderer, {400, 200}, {300, 300}, {300, 0}, {200, 200}, {0, 0, 0});
+    draw_quad_rgb(renderer, {400, 210}, {310, 300}, {600, 300}, {400, 390}, {0, 127, 0});
+    draw_quad_rgb(renderer, {200, 400}, {300, 300}, {300, 600}, {400, 400}, {255, 0, 0});
+    draw_quad_rgb(renderer, {200, 390}, {290, 300}, {0, 300}, {200, 210}, {0, 0, 255});
```
` std::vector<Quad> quads = {
        {{400, 200}, {300, 300}, {300, 0},   {200, 200}, {0, 0, 0}},
        {{400, 210}, {310, 300}, {600, 300}, {400, 390}, {0, 127, 0}},
        {{200, 400}, {300, 300}, {300, 600}, {400, 400}, {255, 0, 0}},
        {{200, 390}, {290, 300}, {0, 300},   {200, 210}, {0, 0, 255}}
    };`

sam-stajnko replied:
I think this would make the code more complicated and would deviate from the original usage example which calls the function 4 times instead:
```cpp
#include "splashkit.h"

int main()
{
    // Create 4 diamond shapes using quads
    quad q1 = quad_from(400, 200, 300, 300, 300, 0, 200, 200);
    quad q2 = quad_from(400, 210, 310, 300, 600, 300, 400, 390);
    quad q3 = quad_from(200, 400, 300, 300, 300, 600, 400, 400);
    quad q4 = quad_from(200, 390, 290, 300, 0, 300, 200, 210);

    open_window("Ninja Star", 600, 600);
    clear_screen(COLOR_WHITE);

    // Draw the quads
    draw_quad(COLOR_BLACK, q1);
    draw_quad(COLOR_GREEN, q2);
    draw_quad(COLOR_RED, q3);
    draw_quad(COLOR_BLUE, q4);

    refresh_screen();
    delay(5000);
    close_all_windows();
    return 0;
}
```

prb616 commented on public/usage-examples/graphics/draw_quad-1-example-beyond.cpp:
```diff
@@ -0,0 +1,101 @@
+};
+
+void draw_quad_rgb(SDL_Renderer * renderer, Point_2D pt_1, Point_2D pt_2, Point_2D pt_3, Point_2D pt_4, Color_RGB color)
+{
+    // Set the line color
```
It takes a single Quad struct as input for cleaner code and future flexibility:
`void draw_quad_rgb(SDL_Renderer* renderer, const Quad& quad)
`

prb616 commented on public/usage-examples/graphics/draw_quad-1-example-beyond.cpp:
```diff
@@ -0,0 +1,101 @@
+    // Render the frame onto the window
+    SDL_RenderPresent(renderer);
+
+    // Wait for 5 seconds before exiting
+    SDL_Delay(5000);
```
The program ended after a 5-second. But it is a suggestion that window should not close automatically after 5sec. So you can add program by which the window will not close until user closes it manually.

`SDL_Event event;
bool running = true;
while (running)
{
    while (SDL_PollEvent(&event))
    {
        if (event.type == SDL_QUIT)
            running = false;
    }
    SDL_Delay(16); // Frame throttling
}
`

sam-stajnko replied:
The reason why the program waits 5 seconds before exiting is because that is what the original usage example does:

```cpp
#include "splashkit.h"

int main()
{
    // Create 4 diamond shapes using quads
    quad q1 = quad_from(400, 200, 300, 300, 300, 0, 200, 200);
    quad q2 = quad_from(400, 210, 310, 300, 600, 300, 400, 390);
    quad q3 = quad_from(200, 400, 300, 300, 300, 600, 400, 400);
    quad q4 = quad_from(200, 390, 290, 300, 0, 300, 200, 210);

    open_window("Ninja Star", 600, 600);
    clear_screen(COLOR_WHITE);

    // Draw the quads
    draw_quad(COLOR_BLACK, q1);
    draw_quad(COLOR_GREEN, q2);
    draw_quad(COLOR_RED, q3);
    draw_quad(COLOR_BLUE, q4);

    refresh_screen();
    delay(5000);
    close_all_windows();
    return 0;
}
```

Since the purpose of the Beyond Splashkit usage examples is to reproduce the original usage example as closely as possible using SDL2, I will keep this the way it is to stay faithful to the original.

prb616 reviewed (changes requested):
I reviewed your code, and I have a few suggestions to improve the structure, efficiency, and scalability of the program:

- [ ] Store quads in a vector and use a loop: Instead of manually calling draw_quad_rgb multiple times, we can define all the quads in a std::vector<Quad> and iterate through them. This eliminates redundancy and makes the code scalable if more quads are added in the future.
- [ ] Update draw_quad_rgb to accept a Quad object: This change complements the use of a vector and allows for cleaner function calls.
- [ ] Replace SDL_Delay(5000) with an event loop: Using an SDL_Event loop allows the user to interact with and close the window when they're ready, making the program more user-friendly.

These changes follow best practices for maintainable and readable C++ code, especially when working with graphical programs. Let me know if you'd like help applying them.

prb616 reviewed (changes requested):
Also please add PR description as per the template.

sam-stajnko commented:
Hi @prb616,
I have committed some changes to my previous commit, so feel free to take another look and tell me what you think. Not all requested changes were acted upon because I felt they would ruin the resemblance to the original usage example for draw_quad. I'm not sure if you're aware but the purpose of the Beyond Splashkit usage examples is to reproduce the previous usage example but without using the Splashkit core library. This means that the structure and functionality of the code should still resemble the original usage example as closely as possible. Also the code should be as simple as possible and SDL functions and data structures should be prioritized over writing it from scratch as much as possible.

I have left comments on your individual recommendations as to why I disagree so take a look at these.

With that being said, I have implemented a couple of changes to my code, some of which are your recommendations and others that are my own:

- I have created a struct named Quad.
- The Quad struct is now passed as an argument to the DrawQuad function instead of 4 points.
- SDL_Point and SDL_Color are now used instead of my own data structures.
- draw_quad_rgb has been renamed to just draw_quad.
- The create_window function has been removed. All window creation now takes place in main to reduce the number of functions.
- A function called quad_from has been created that returns a Quad from four points. This has been added because such a function is used in the original usage example.
- Comments have been minimized and simplified where possible to make the code more readable and cleaner.
- I have added a description to the pull request.

huyhuynh2319 reviewed (approved):
I've reviewed the usage examples. Everything ran without issue.
# **Checks**

- [x]All required files are present.
- [x]C++ code
- [x]Code clearly demonstrates the function.
- [x]All versions maintain the same structure and comments.
- [x]C++ code ran correctly.

# **Tests done**

- [x]  C++ Beyond SplashKit code ran correctly.
